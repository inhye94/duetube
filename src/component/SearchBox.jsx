import { debounce } from "lodash";
import React, { useEffect, useState } from "react";
import { IoMdClose } from "react-icons/io";
import { IoSearch } from "react-icons/io5";
import IconButton from "../modules/IconButton";
import SearchForm from "./SearchForm";

export default function SearchBox() {
  const [_isMobile, setMobile] = useState(checkMobile());
  const [_searchToggle, setSearchToggle] = useState(!checkMobile());

  const handleResize = debounce(() => {
    setMobile(window.innerWidth < 768 ? true : false);
  }, 500);

  useEffect(() => {
    window.addEventListener("load", handleResize);
    window.addEventListener("resize", handleResize);
    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, [handleResize]);

  useEffect(() => {
    setSearchToggle(_isMobile ? false : true);
  }, [_isMobile]);

  return (
    <>
      {_isMobile && (
        <IconButton text="검색" callback={() => setSearchToggle(true)}>
          <IoSearch aria-label="검색" />
        </IconButton>
      )}

      <div
        className={`flex items-center fixed left-0 z-10 w-full px-[8px] ${
          _searchToggle ? "translate-y-[0]" : "translate-y-[-64px]"
        }  transition-transform duration-300 bg-white md:static md:min-w-[480px] md:px-0 md:bg-transparent dark:bg-slate-800`}
      >
        {_isMobile && (
          <IconButton
            text="검색창 닫기"
            extraStyle={"mr-[4px]"}
            callback={() => setSearchToggle(false)}
          >
            <IoMdClose aria-label="검색창 닫기" />
          </IconButton>
        )}

        <SearchForm blurEvent={() => _isMobile && setSearchToggle(false)} />
      </div>
    </>
  );
}

const checkMobile = () => {
  return window.innerWidth < 768 ? true : false;
};
