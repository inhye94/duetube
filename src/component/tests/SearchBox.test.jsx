import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import SearchBox from "../SearchBox";

const SHOW_ELEMENT = "translate-y-[0]";
const ORIGINAL_INNER_WIDTH = window.innerWidth;

// 윈도우 크기를 조정하는 함수
function resizeWindow(width) {
  Object.defineProperty(window, "innerWidth", {
    writable: true,
    configurable: true,
    value: width,
  });
  window.dispatchEvent(new Event("resize"));
}

// SearchForm mocking
jest.mock("../SearchForm", () => {
  return (props) => (
    <form data-testid="search-form" onBlur={() => props.blurEvent?.()}>
      <input type="text" data-testid="search-input" />
    </form>
  );
});

describe("SearchBox", () => {
  it("초기 렌더 - PC", () => {
    resizeWindow(1200);
    const { asFragment } = render(<SearchBox />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("초기 렌더 - 모바일", () => {
    resizeWindow(300);
    const { asFragment } = render(<SearchBox />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("모바일 동작", () => {
    let openButton, wrapper, closeButton;

    beforeEach(() => {
      resizeWindow(300);
      render(<SearchBox />);

      openButton = screen.getByLabelText("검색창 열기");
      wrapper = screen.getByTestId("search-form").parentElement;
      closeButton = screen.getByLabelText("검색창 닫기");
    });

    it("검색 버튼을 클릭하면 wrapper가 열림", () => {
      fireEvent.click(openButton);
      expect(wrapper?.className.includes(SHOW_ELEMENT)).toBe(true);
    });

    it("닫기 버튼을 클릭하면 wrapper가 닫힘", () => {
      fireEvent.click(openButton);
      fireEvent.click(closeButton);
      expect(wrapper?.className.includes(SHOW_ELEMENT)).toBe(false);
    });

    it("SearchForm에서 blur 이벤트 발생하면 wrapper가 닫힘", () => {
      const searchInput = screen.getByTestId("search-input");
      fireEvent.click(openButton);
      fireEvent.blur(searchInput);

      expect(wrapper?.className.includes(SHOW_ELEMENT)).toBe(false);
    });

    it("윈도우 리사이즈(모바일 > 데스크탑)되면, wrapper가 닫힘", () => {
      resizeWindow(1200);
      expect(wrapper?.className.includes(SHOW_ELEMENT)).toBe(false);
    });
  });

  afterEach(() => {
    Object.defineProperty(window, "innerWidth", {
      writable: true,
      configurable: true,
      value: ORIGINAL_INNER_WIDTH,
    });
  });
});
