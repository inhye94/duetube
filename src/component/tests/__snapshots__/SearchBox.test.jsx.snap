// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox 초기 렌더 - PC 1`] = `
<DocumentFragment>
  <div
    class="flex items-center fixed left-0 z-10 w-full px-[8px] translate-y-[0]  transition-transform duration-300 bg-white md:static md:min-w-[480px] md:px-0 md:bg-transparent dark:bg-slate-800"
  >
    <form
      data-testid="search-form"
    >
      <input
        data-testid="search-input"
        type="text"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`SearchBox 초기 렌더 - 모바일 1`] = `
<DocumentFragment>
  <button
    aria-label="검색창 열기"
    class="shrink-0 flex justify-center items-center w-8 h-8 text-[20px] rounded-full transition-colors hover:bg-slate-200 dark:hover:bg-slate-600"
    title="검색"
    type="button"
  >
    <svg
      aria-label="검색"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M456.69 421.39L362.6 327.3a173.81 173.81 0 0034.84-104.58C397.44 126.38 319.06 48 222.72 48S48 126.38 48 222.72s78.38 174.72 174.72 174.72A173.81 173.81 0 00327.3 362.6l94.09 94.09a25 25 0 0035.3-35.3zM97.92 222.72a124.8 124.8 0 11124.8 124.8 124.95 124.95 0 01-124.8-124.8z"
      />
    </svg>
  </button>
  <div
    class="flex items-center fixed left-0 z-10 w-full px-[8px] translate-y-[-64px]  transition-transform duration-300 bg-white md:static md:min-w-[480px] md:px-0 md:bg-transparent dark:bg-slate-800"
  >
    <button
      aria-label="검색창 닫기"
      class="shrink-0 flex justify-center items-center w-8 h-8 text-[20px] rounded-full transition-colors hover:bg-slate-200 dark:hover:bg-slate-600 mr-[4px]"
      title="검색창 닫기"
      type="button"
    >
      <svg
        aria-label="닫기"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
        />
      </svg>
    </button>
    <form
      data-testid="search-form"
    >
      <input
        data-testid="search-input"
        type="text"
      />
    </form>
  </div>
</DocumentFragment>
`;
